### PR TITLE
CTC loss for TF and torch

### PR DESCRIPTION
**Pull request by MaanasArora**
Part of the CTC API for TensorFlow and torch. I'd like to start on the JAX implementation after this.

It can be optimized with further work, so please just let me know any general points I would need to address and I can optimize, refactor, and write the unit tests. Thank you!

### CODE REVIEW

File: keras/backend/tensorflow/nn.py [hunk 1]
```diff
@@ -787,3 +787,69 @@ def batch_normalization(
...
+    Returns:
+        A tensor of shape `(batch_size,)` containing the CTC loss for each
+        sample in the batch.
+    """
+    target = tf.convert_to_tensor(target)
+    target = tf.cast(target, dtype="int32")
+    output = tf.convert_to_tensor(output)
+    output = tf.cast(output, dtype="float32")
+
+    sparse_target = ctc_labels_to_sparse(target, target_length)
```
Comment (fchollet) [inline keras/backend/tensorflow/nn.py:847]: You can just inline the function's contents here.

File: keras/backend/tensorflow/nn.py [hunk 2]
```diff
@@ -787,3 +787,69 @@ def batch_normalization(
...
+    target_length,
+    output_length,
+    blank_index=0,
+):
+    """Runs CTC (Connectionist Temporal Classification) loss on each batch element.
+
+    Arguments:
+        target: Tensor `(batch_size, max_target_length)` containing the
+            target sequences.
+        output: Tensor `(batch_size, max_output_length, label_classes)`
```
Comment (fchollet) [inline keras/backend/tensorflow/nn.py:829]: label_classes -> num_classes

File: keras/backend/tensorflow/nn.py [hunk 3]
```diff
@@ -787,3 +787,69 @@ def batch_normalization(
...
+    output,
+    target_length,
+    output_length,
+    blank_index=0,
+):
+    """Runs CTC (Connectionist Temporal Classification) loss on each batch element.
+
+    Arguments:
+        target: Tensor `(batch_size, max_target_length)` containing the
+            target sequences.
```
Comment (fchollet) [inline keras/backend/tensorflow/nn.py:828]: in integer format

File: keras/backend/tensorflow/nn.py [hunk 4]
```diff
@@ -787,3 +787,69 @@ def batch_normalization(
...
+        output: Tensor `(batch_size, max_output_length, label_classes)`
+            containing the output of the softmax.
+        target_length: Tensor `(batch_size,)` containing the sequence length
+            for each target sequence in the batch.
+        output_length: Tensor `(batch_size,)` containing the sequence length
+            for each output sequence in the batch.
+        blank_index: The value in `target` and `output` that represents the
```
Comment (fchollet) [inline keras/backend/tensorflow/nn.py:835]: Is there a name that would be more consistent with naming conventions elsewhere in the API, like `padding_index`?

**Review by fchollet — COMMENTED** [review]
Thank you for the PR! This is looking good.

File: keras/backend/tensorflow/nn.py (hunk 3 above)
Comment (MaanasArora) [reply keras/backend/tensorflow/nn.py:828]: Done, thanks!

File: keras/backend/tensorflow/nn.py (hunk 2 above)
Comment (MaanasArora) [reply keras/backend/tensorflow/nn.py:829]: Yes, that's consistent. Thanks!

File: keras/backend/tensorflow/nn.py (hunk 4 above)
Comment (MaanasArora) [reply keras/backend/tensorflow/nn.py:835]: Blanks can actually occur in between two "actual" labels and are not necessarily padding, so it might not fit the CTC spec (which explicitly calls these blanks). Should we stick to the CTC name? It seems quite specific.

File: keras/backend/tensorflow/nn.py (hunk 1 above)
Comment (MaanasArora) [reply keras/backend/tensorflow/nn.py:847]: Done, thanks.

File: keras/backend/tensorflow/nn.py (hunk 4 above)
Comment (fchollet) [reply keras/backend/tensorflow/nn.py:810]: Sure, that is fine!

**Comment by MaanasArora** [comment]
I addressed the review and just fixed a formatting issue that was failing the check. This should be ready!

**Comment by MaanasArora** [comment]
Sorry--fixed one more small linting issue (trailing space).

File: keras/backend/tensorflow/nn.py [hunk 5]
```diff
@@ -787,3 +787,55 @@ def batch_normalization(
         scale=scale,
         variance_epsilon=epsilon,
     )
+
+
+def ctc_batch_cost(
```
Comment (fchollet) [inline keras/backend/tensorflow/nn.py:792]: Let's rename to `ctc_loss` to stay in line with `tf.nn` and `torch.nn`.

**Review by fchollet — COMMENTED** [review]
Thanks for the update! Please add an op in `keras/ops/nn.py` and add correctness unit tests (this is important in particular to check that we have the same numerics across different backends).

File: keras/backend/tensorflow/nn.py (hunk 5 above)
Comment (MaanasArora) [reply keras/backend/tensorflow/nn.py]: Done, thanks!

File: keras/backend/tensorflow/nn.py (hunk 4 above)
Comment (MaanasArora) [reply keras/backend/tensorflow/nn.py]: Sure, this is done!

File: keras/ops/nn.py [hunk 6]
```diff
@@ -1777,3 +1777,61 @@ def batch_normalization(
...
+            "target_length", target_length.shape, "target", target.shape
+        )
+        self._check_shape_first_dim(
+            "output_length", output_length.shape, "output", output.shape
+        )
+
+        return KerasTensor((target.shape[0],), dtype=target.dtype)
+
+
+@keras_export("keras.ops.ctc_loss")
```
Comment (fchollet) [inline keras/ops/nn.py:1813]: You can also export it to `keras.ops.nn.ctc_loss`

File: keras/ops/nn_test.py [hunk 7]
```diff
@@ -1750,6 +1750,22 @@ def test_batch_normalization(self):
         )
         self.assertEqual(tuple(output.shape), (2, 3, 3, 5))
 
+    def test_ctc_loss(self):
```
Comment (fchollet) [inline keras/ops/nn_test.py:1753]: Please also add a shape inference test (calling the function on KerasTensors and checking the shape/dtype of the output) -- see the TestCase class above

Comment (fchollet) [inline keras/ops/nn_test.py:1753]: Either skip the JAX backend in this test, or implement the ctc_loss op for JAX as well

**Review by fchollet — COMMENTED** [review]
Thank you for the update!

File: keras/ops/nn_test.py (hunk 7 above)
Comment (MaanasArora) [reply keras/ops/nn_test.py:1768]: Done, thank you!

File: keras/ops/nn.py (hunk 6 above)
Comment (MaanasArora) [reply keras/ops/nn.py]: Done

File: keras/ops/nn_test.py (hunk 7 above)
Comment (MaanasArora) [reply keras/ops/nn_test.py:1768]: I skipped it for now, thanks

File: keras/ops/nn_test.py [hunk 8]
```diff
@@ -974,6 +974,17 @@ def test_batch_normalization(self):
             (10, 3, 4, 5),
         )
 
+    @pytest.mark.skipif(
+        backend.backend() == "jax",
```
Comment (fchollet) [inline keras/ops/nn_test.py:978]: Rather, use `backend.backend() not in ["tensorflow", "torch"]` since it is not implemented in the NumPy backend either

Comment (MaanasArora) [reply keras/ops/nn_test.py]: Oh I overlooked that, fixed, thanks.

**Comment by MaanasArora** [comment]
I just fixed the linting issue as well, so that no tests fail

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the great contribution!
